// accord/lib/adapters/coco/0.9.x.js
// Generated by CoffeeScript 1.12.7
(function() {
  var Adapter, Coco, W,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Adapter = require('../../adapter_base');

  W = require('when');

  Coco = (function(superClass) {
    var compile;

    extend(Coco, superClass);

    function Coco() {
      return Coco.__super__.constructor.apply(this, arguments);
    }

    Coco.prototype.name = 'coco';

    Coco.prototype.extensions = ['co'];

    Coco.prototype.output = 'js';

    Coco.prototype.isolated = true;

    Coco.prototype._render = function(str, options) {
      return compile((function(_this) {
        return function() {
          return _this.engine.compile(str, options);
        };
      })(this));
    };

    compile = function(fn) {
      var err, res;
      try {
        res = fn();
      } catch (error) {
        err = error;
        return W.reject(err);
      }
      return W.resolve({
        result: res
      });
    };

    return Coco;

  })(Adapter);

  module.exports = Coco;

}).call(this);
