// accord/lib/adapters/dot/1.x.js
// Generated by CoffeeScript 1.12.7
(function() {
  var Adapter, Dot, W, fs, path,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Adapter = require('../../adapter_base');

  path = require('path');

  fs = require('fs');

  W = require('when');

  Dot = (function(superClass) {
    var compile;

    extend(Dot, superClass);

    function Dot() {
      return Dot.__super__.constructor.apply(this, arguments);
    }

    Dot.prototype.name = 'dot';

    Dot.prototype.extensions = ['dot'];

    Dot.prototype.output = 'html';

    Dot.prototype._render = function(str, options) {
      return compile((function(_this) {
        return function() {
          return _this.engine.compile(str)(options);
        };
      })(this));
    };

    Dot.prototype._compile = function(str, options) {
      return compile((function(_this) {
        return function() {
          return _this.engine.compile(str, options);
        };
      })(this));
    };

    Dot.prototype._compileClient = function(str, options) {
      options.client = true;
      return compile((function(_this) {
        return function() {
          return _this.engine.compile(str, options).toString();
        };
      })(this));
    };

    Dot.prototype.clientHelpers = function(str, options) {
      var runtime_path;
      runtime_path = path.join(this.engine.__accord_path, 'doT.min.js');
      return fs.readFileSync(runtime_path, 'utf8');
    };

    compile = function(fn) {
      var err, res;
      try {
        res = fn();
      } catch (error) {
        err = error;
        return W.reject(err);
      }
      return W.resolve({
        result: res
      });
    };

    return Dot;

  })(Adapter);

  module.exports = Dot;

}).call(this);
