// accord/lib/adapters/babel/4.x - 5.x.js
// Generated by CoffeeScript 1.12.7
(function() {
  var Adapter, Babel, W, path, sourcemaps,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Adapter = require('../../adapter_base');

  path = require('path');

  W = require('when');

  sourcemaps = require('../../sourcemaps');

  Babel = (function(superClass) {
    var compile;

    extend(Babel, superClass);

    function Babel() {
      return Babel.__super__.constructor.apply(this, arguments);
    }

    Babel.prototype.name = 'babel';

    Babel.prototype.extensions = ['jsx', 'js'];

    Babel.prototype.output = 'js';

    Babel.prototype.isolated = true;

    Babel.prototype._render = function(str, options) {
      var filename;
      filename = options.filename;
      if (options.sourcemap === true) {
        options.sourceMap = true;
      }
      options.sourceMapName = filename;
      delete options.sourcemap;
      return compile((function(_this) {
        return function() {
          return _this.engine.transform(str, options);
        };
      })(this));
    };

    compile = function(fn) {
      var data, err, res;
      try {
        res = fn();
      } catch (error) {
        err = error;
        return W.reject(err);
      }
      data = {
        result: res.code
      };
      if (res.map) {
        return sourcemaps.inline_sources(res.map).then(function(map) {
          data.sourcemap = map;
          return data;
        });
      } else {
        return W.resolve(data);
      }
    };

    return Babel;

  })(Adapter);

  module.exports = Babel;

}).call(this);
