// accord/lib/adapters/dogescript/2.x.js
// Generated by CoffeeScript 1.12.7
(function() {
  var Adapter, DogeScript, W,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Adapter = require('../../adapter_base');

  W = require('when');

  DogeScript = (function(superClass) {
    extend(DogeScript, superClass);

    function DogeScript() {
      return DogeScript.__super__.constructor.apply(this, arguments);
    }

    DogeScript.prototype.name = 'dogescript';

    DogeScript.prototype.extensions = ['djs'];

    DogeScript.prototype.output = 'js';

    DogeScript.prototype.isolated = true;

    DogeScript.prototype._render = function(str, options) {
      return W.resolve({
        result: this.engine(str, options.beauty, options.trueDoge)
      });
    };

    return DogeScript;

  })(Adapter);

  module.exports = DogeScript;

}).call(this);
