// accord/lib/adapters/pug/2.0.0-rc - 2.x.js
// Generated by CoffeeScript 1.12.7
(function() {
  var Adapter, Pug, UglifyJS, W, fs, path,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Adapter = require('../../adapter_base');

  path = require('path');

  fs = require('fs');

  W = require('when');

  UglifyJS = require('uglify-js');

  Pug = (function(superClass) {
    var compile;

    extend(Pug, superClass);

    function Pug() {
      return Pug.__super__.constructor.apply(this, arguments);
    }

    Pug.prototype.name = 'pug';

    Pug.prototype.extensions = ['pug'];

    Pug.prototype.output = 'html';

    Pug.prototype.supportedEngines = ['pug'];

    Pug.prototype._render = function(str, options) {
      return compile((function(_this) {
        return function() {
          return _this.engine.render(str, options);
        };
      })(this));
    };

    Pug.prototype._compile = function(str, options) {
      return compile((function(_this) {
        return function() {
          return _this.engine.compile(str, options);
        };
      })(this));
    };

    Pug.prototype._compileClient = function(str, options) {
      return compile((function(_this) {
        return function() {
          return _this.engine.compileClient(str, options);
        };
      })(this));
    };

    compile = function(fn) {
      var err, res;
      try {
        res = fn();
      } catch (error) {
        err = error;
        return W.reject(err);
      }
      return W.resolve({
        result: res
      });
    };

    return Pug;

  })(Adapter);

  module.exports = Pug;

}).call(this);
