// accord/lib/adapters/haml/0.6.x.js
// Generated by CoffeeScript 1.12.7
(function() {
  var Adapter, HAML, UglifyJS, W, fs, path,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Adapter = require('../../adapter_base');

  path = require('path');

  fs = require('fs');

  W = require('when');

  UglifyJS = require('uglify-js');

  HAML = (function(superClass) {
    var compile;

    extend(HAML, superClass);

    function HAML() {
      return HAML.__super__.constructor.apply(this, arguments);
    }

    HAML.prototype.name = 'haml';

    HAML.prototype.extensions = ['haml'];

    HAML.prototype.output = 'html';

    HAML.prototype.supportedEngines = ['hamljs'];

    HAML.prototype._render = function(str, options) {
      return compile((function(_this) {
        return function() {
          return _this.engine.compile(str)(options);
        };
      })(this));
    };

    HAML.prototype._compile = function(str, options) {
      return compile((function(_this) {
        return function() {
          return _this.engine.compile(str, options);
        };
      })(this));
    };

    compile = function(fn) {
      var err, res;
      try {
        res = fn();
      } catch (error) {
        err = error;
        return W.reject(err);
      }
      return W.resolve({
        result: res
      });
    };

    return HAML;

  })(Adapter);

  module.exports = HAML;

}).call(this);
